File: src/main/java/com/airwaves/airwavesweb/tasks/ClusterController.java
package com.airwaves.airwavesweb.tasks;

import com.airwaves.airwavesweb.datastore.Cluster;
import com.airwaves.airwavesweb.datastore.User;
import com.airwaves.airwavesweb.util.LocationWrapper;
import org.apache.commons.math3.ml.clustering.KMeansPlusPlusClusterer;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@RestController
public class ClusterController {

    @GetMapping(value = "/update-clusters")
    public void update_clusters() {
        var oldClusters = Cluster.getAll();
        for (var cluster : oldClusters) {
            cluster.delete();
        }

        List<LocationWrapper> users = new ArrayList<>();
        for (User user : User.getAll()) {
            users.add(new LocationWrapper(user));
        }

        var clusterer = new KMeansPlusPlusClusterer<LocationWrapper>(users.size() / Cluster.maxUsers);
        var results = clusterer.cluster(users);

        for (var result : results) {
            var clusteredUsers = result.getPoints().stream().map(LocationWrapper::getUser).collect(Collectors.toList());
            var cluster = new Cluster(clusteredUsers, result.getCenter().getPoint()[0], result.getCenter().getPoint()[1]);
            for (var user : clusteredUsers) {
                user.setCluster(cluster);
            }
        }
    }
}

File: src/main/kotlin/com/airwaves/airwavesweb/tasks/ClusterController.kt
package com.airwaves.airwavesweb.tasks

import com.airwaves.airwavesweb.datastore.Cluster
import com.airwaves.airwavesweb.datastore.Cluster.Companion.maxUsers
import com.airwaves.airwavesweb.datastore.User
import com.airwaves.airwavesweb.util.LocationWrapper
import org.apache.commons.math3.ml.clustering.KMeansPlusPlusClusterer
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RestController
import java.util.*

@RestController
class ClusterController {
    @GetMapping("/update-clusters")
    fun update_clusters() {
        val oldClusters = Cluster.all
        for (cluster in oldClusters) {
            cluster.delete()
        }
        val users: MutableList<LocationWrapper> = ArrayList()
        for (user in User.all) {
            users.add(LocationWrapper(user))
        }
        val clusterer = KMeansPlusPlusClusterer<LocationWrapper>(users.size / maxUsers)
        val results = clusterer.cluster(users)
        for (result in results) {
            val clusteredUsers = result.points.map { x -> x.user }
            val cluster = Cluster(null, hashMapOf(
                    "users" to clusteredUsers.size.toLong(),
                    "latitude" to result.center.point[0],
                    "longitude" to result.center.point[1]
            ))
            cluster.save()
            for (user in clusteredUsers) {
                user.cluster = cluster
                user.save()
            }
        }
    }
}